README
Functions for Improving Accuracy of School-Level Aggregate Growth Measures Using Empirical Best Linear Prediction


NEWS
###################
## TODOs
###################

## 10/28/2021: bugfix:
update maxit to 100,000 in the matrix-utils.R file for the Radj function:
.res <- nearPD2(Madj, fix0s=TRUE, do2eigen=FALSE, eig.tol=eig.tol, conv.tol=1e-11, maxit=100000_)

allow target to be a list of named vectors so that we get BLPs for a vector of
outcomes simultaneously

generalize the syntax of target to allow importance weights on blocks, like you
might need if defining a multi-year target

add something about variance decomposition diagnostics (comparing estimated and
observed moments), perhaps after adjusting for block*pattern effects

create S3 class and associated summary and plot methods


R/blp.R
blp <- function(Y, lambda, mu, SigmaX, SigmaU, eig.tol, checkMSE=FALSE){
    ## X has E[X] = mu, var[X] = SigmaX
    ## Y = X + U where E[U|X] = 0, var[U] = SigmaU
    ## compute BLP of lambda'X from Y, along with MSE and PRMSE, conditional
    ## on mu being fixed and known.
    ##
    ## algebra:
    ## Q = SigmaX(SigmaX + SigmaU)^{-1}
    ## BLP = lambda'[ (I-Q)mu + QY ]
    ## MSE = tr(lambda lambda' (Q SigmaU Q' + (I-Q)SigmaX(I-Q)')) = lambda' (I-Q)SigmaX lambda
    ##
    ## NOTE: all eigenvalues of (SigmaX + SigmaU) that are less than eig.tol*max eigenvalue
    ## are zeroed, and the Moore-Penrose inverse is used to get coefficients for BLP.

    ## checks
    stopifnot(is.matrix(SigmaX) && is.matrix(SigmaU))    
    stopifnot(is.numeric(Y) && is.numeric(lambda) && is.numeric(mu) && is.numeric(SigmaX) && is.numeric(SigmaU) && is.numeric(eig.tol))
    stopifnot(all(!is.na(c(Y,lambda,mu,SigmaX,SigmaU,eig.tol))))
    p <- length(Y)
    if(p==1){
        dim(SigmaX) <- dim(SigmaU) <- c(1,1)
    }
    stopifnot( (length(lambda) == p) && (length(mu) == p) && (nrow(SigmaX) == p) && (ncol(SigmaX) == p) && (nrow(SigmaU) == p) && (ncol(SigmaU) == p) )
    stopifnot(max(abs(SigmaX - t(SigmaX))) < 1e-8)
    stopifnot(max(abs(SigmaU - t(SigmaU))) < 1e-8)
    dim(Y) <- dim(lambda) <- dim(mu) <-  c(p,1)

    ## adjustment based on eigenvalues
    SigmaXpU <- SigmaX + SigmaU
    e        <- eigen(SigmaXpU)
    tozero   <- which(e$values < max(abs(e$values))*eig.tol)
    if(length(tozero) > 0){
        print("warning in blp: eigenvalues of SigmaX + SigmaU adjusted")
        e$values[tozero] <- 0.0
        SigmaXpU <- e$vectors %*% diag(e$values) %*% t(e$vectors)
    }
    
    ## computations
    est.direct <- as.vector(t(lambda) %*% Y)
    mse.direct <- as.vector(t(lambda) %*% SigmaU %*% lambda)
    Q          <- SigmaX %*% ginv(SigmaXpU, tol = eig.tol)
    if(max(abs( (Q %*% SigmaXpU) - SigmaX)) > 1e-06){
        print("warning in blp: G-inverse solution to BLP coefficients only approximate")
    }
    IminusQ    <- diag(p) - Q
    est.blp    <- as.vector(t(lambda) %*% ( (IminusQ %*% mu) + (Q %*% Y) ))
    mse.blp    <- as.vector(t(lambda) %*% (IminusQ %*% SigmaX) %*% lambda)

    if(checkMSE){
        part       <- (Q %*% SigmaU %*% t(Q)) + (IminusQ %*% SigmaX %*% t(IminusQ))
        mse.blp2    <- sum(diag( (lambda %*% t(lambda)) %*% part))
        stopifnot(abs(mse.blp - mse.blp2) < 1e-10)
    }

    ## PRMSE of BLP relative to assigning everyone the mean lambda'mu
    mse.null   <- as.vector(t(lambda) %*% SigmaX %*% lambda)
    prmse.null <- 1 - mse.blp/mse.null

    ## PRMSE of BLP relative to direct estimator
    prmse.direct <- 1 - mse.blp/mse.direct

    return(list(est     = c(est.direct = est.direct, mse.direct = mse.direct, est.blp = est.blp, mse.blp = mse.blp, prmse.null = prmse.null, prmse.direct = prmse.direct),
                Q       = Q,
                IminusQ = IminusQ))
}

## simple case
##
## library(JRLmisc)
## Y       <- rnorm(6)
## lambda  <- c(1,0,3,2,0,0)
## mu      <- c(0,1,4,0,0,1)
## sdx     <- sqrt(c(0.2, 0.3, 0.5, 0.2, 0.2, 0.3))
## SigmaX  <- diag(sdx) %*% CS(0.3,6) %*% diag(sdx)
## sdu     <- sqrt(c(0.1, 0.1, 0.15, 0.05, 0.10, 0.15))
## SigmaU  <- diag(sdu) %*% CS(0.1,6) %*% diag(sdu)
## blp(Y, lambda, mu, SigmaX, SigmaU, checkMSE=TRUE)
##
## case where eigenvalues get adjusted
##
## e <- eigen(SigmaX)
## e$values[which(e$values < 0.20)] <- -0.10
## SigmaX <- e$vectors %*% diag(e$values) %*% t(e$vectors)
## blp(Y, lambda, mu, SigmaX, SigmaU, checkMSE=TRUE)



R/nearPD2.R
nearPD2	<- function (x,fix0s=FALSE,corr = FALSE, keepDiag = FALSE, do2eigen = TRUE, 
    doSym = FALSE, doDykstra = TRUE, only.values = FALSE, ensureSymmetry = !isSymmetric(x), 
    eig.tol = 1e-06, conv.tol = 1e-07, posd.tol = 1e-08, maxit = 100, 
    conv.norm.type = "I", trace = FALSE) 
{
    if(fix0s && !is(x,"sparseMatrix")){
        stop("fix0s option valid only when x is a sparse matrix")
    }
    if (ensureSymmetry) {
        x <- symmpart(x)
    }
    n <- ncol(x)
    if (keepDiag) 
        diagX0 <- diag(x)
    if (doDykstra) {
        D_S <- x
        D_S[] <- 0
    }
    X <- x
    iter <- 0
    converged <- FALSE
    conv <- Inf
    while (iter < maxit && !converged) {
        Y <- X
        if (doDykstra) 
            R <- Y - D_S
        e <- eigen(if (doDykstra) 
            R
        else Y, symmetric = TRUE)
        Q <- e$vectors
        d <- e$values
        p <- d > eig.tol * d[1]
        if (!any(p)) 
            stop("Matrix seems negative semi-definite")
        Q <- Q[, p, drop = FALSE]
        X <- tcrossprod(Q * rep(d[p], each = nrow(Q)), Q)
        if (doDykstra) 
            D_S <- X - R
        if (doSym) 
            X <- (X + t(X))/2
        if (corr) 
            diag(X) <- 1
        else if (keepDiag) 
            diag(X) <- diagX0
##added
##	X[which0]	<- 0
	if(fix0s==TRUE) {
		vind		<- data.frame(i=x@i)
		vind$Row	<- rep(1:nrow(x),diff(x@p))
		vind$Col	<- vind$i+1
		vind$ind	<- (vind$Col-1)*nrow(x)+vind$Row
		X		<- sparseMatrix(i=vind$Row,j=vind$Col,x=X[vind$ind],symmetric=TRUE,dims=c(nrow(X),ncol(X)))
	}

        conv <- norm(Y - X, conv.norm.type)/norm(Y, conv.norm.type)
        iter <- iter + 1
        if (trace) 
            cat(sprintf("iter %3d : #{p}=%d, ||Y-X|| / ||Y||= %11g\n", 
                iter, sum(p), conv))
        converged <- (conv <= conv.tol)
    }
    if (!converged) {
        warning(gettextf("'nearPD()' did not converge in %d iterations", 
                         iter), domain = NA)
        return(list(converged=FALSE))
    } else {
        if (do2eigen || only.values) {
            e <- eigen(X, symmetric = TRUE)
            d <- e$values
            Eps <- posd.tol * abs(d[1])
            if (d[n] < Eps) {
                d[d < Eps] <- Eps
                if (!only.values) {
                    Q <- e$vectors
                    o.diag <- diag(X)
                    X <- Q %*% (d * t(Q))
                    D <- sqrt(pmax(Eps, o.diag)/diag(X))
                    X[] <- D * X * rep(D, each = n)
                }
            }
            if (only.values) 
                return(d)
            if (corr) 
                diag(X) <- 1
            else if (keepDiag) 
                diag(X) <- diagX0
        }
        structure(list(mat = new("dpoMatrix", x = as.vector(X), Dim = c(n, 
                                                                        n)), 
                       ##	Dimnames = .M.DN(x)), ##commented out 
                       eigenvalues = d, corr = corr, 
                       normF = norm(x - X, "F"), iterations = iter, rel.tol = conv, 
                       converged = converged), class = "nearPD")
    }
}


R/matrix-utils.R
nearPD2	<- function (x,fix0s=FALSE,corr = FALSE, keepDiag = FALSE, do2eigen = TRUE, 
    doSym = FALSE, doDykstra = TRUE, only.values = FALSE, ensureSymmetry = !isSymmetric(x), 
    eig.tol = 1e-06, conv.tol = 1e-07, posd.tol = 1e-08, maxit = 100, 
    conv.norm.type = "I", trace = FALSE)
{
    ## adaptation of nearPD() from library(Matrix) to enforce fixed zeros via 'fix0s' argument
    if(fix0s && !is(x,"sparseMatrix")){
        stop("fix0s option valid only when x is a sparse matrix")
    }
    if (ensureSymmetry) {
        x <- symmpart(x)
    }
    n <- ncol(x)
    if (keepDiag) 
        diagX0 <- diag(x)
    if (doDykstra) {
        D_S <- x
        D_S[] <- 0
    }
    X <- x
    iter <- 0
    converged <- FALSE
    conv <- Inf
    while (iter < maxit && !converged) {
        Y <- X
        if (doDykstra) 
            R <- Y - D_S
        e <- eigen(if (doDykstra) 
            R
        else Y, symmetric = TRUE)
        Q <- e$vectors
        d <- e$values
        p <- d > eig.tol * d[1]
        if (!any(p)) 
            stop("Matrix seems negative semi-definite")
        Q <- Q[, p, drop = FALSE]
        X <- tcrossprod(Q * rep(d[p], each = nrow(Q)), Q)
        if (doDykstra) 
            D_S <- X - R
        if (doSym) 
            X <- (X + t(X))/2
        if (corr) 
            diag(X) <- 1
        else if (keepDiag) 
            diag(X) <- diagX0
##added
##	X[which0]	<- 0
	if(fix0s==TRUE) {
		vind		<- data.frame(i=x@i)
		vind$Row	<- rep(1:nrow(x),diff(x@p))
		vind$Col	<- vind$i+1
		vind$ind	<- (vind$Col-1)*nrow(x)+vind$Row
		X		<- sparseMatrix(i=vind$Row,j=vind$Col,x=X[vind$ind],symmetric=TRUE,dims=c(nrow(X),ncol(X)))
	}

        conv <- norm(Y - X, conv.norm.type)/norm(Y, conv.norm.type)
        iter <- iter + 1
        if (trace) 
            cat(sprintf("iter %3d : #{p}=%d, ||Y-X|| / ||Y||= %11g\n", 
                iter, sum(p), conv))
        converged <- (conv <= conv.tol)
    }
    if (!converged) {
        warning(gettextf("'nearPD()' did not converge in %d iterations", 
                         iter), domain = NA)
        return(list(converged=FALSE))
    } else {
        if (do2eigen || only.values) {
            e <- eigen(X, symmetric = TRUE)
            d <- e$values
            Eps <- posd.tol * abs(d[1])
            if (d[n] < Eps) {
                d[d < Eps] <- Eps
                if (!only.values) {
                    Q <- e$vectors
                    o.diag <- diag(X)
                    X <- Q %*% (d * t(Q))
                    D <- sqrt(pmax(Eps, o.diag)/diag(X))
                    X[] <- D * X * rep(D, each = n)
                }
            }
            if (only.values) 
                return(d)
            if (corr) 
                diag(X) <- 1
            else if (keepDiag) 
                diag(X) <- diagX0
        }
        structure(list(mat = new("dpoMatrix", x = as.vector(X), Dim = c(n, 
                                                                        n)), 
                       ##	Dimnames = .M.DN(x)), ##commented out 
                       eigenvalues = d, corr = corr, 
                       normF = norm(x - X, "F"), iterations = iter, rel.tol = conv, 
                       converged = converged), class = "nearPD")
    }
}


Radj <- function(M, nzlocs, adj_method, eig.tol, eig.min=NULL){
    ## function for R adjustment to PSD, maintaining fixed zeros
    ## if specified via allowing only the non-zero elements in "nzlocs"
    ##
    ## if adj_method=="nearPD", calls nearPD2 with eig.tol
    ##
    ## if adj_method=="spectral", iteratively adjusts eigenvalues that are
    ## less than (eig.tol * max) and imposing fixed zeros until
    ## all eigenvalues exceed (eig.min * max)
    ##
    ## NOTE: no sum-to-zero constraints in this case
    ## NOTE: when adj_method=="nearPD" there should be no need for iteration,
    ## although for simplificity of code, it is placed within the while() loop
    stopifnot( (class(M) == "dsCMatrix") && is.matrix(nzlocs) && is.character(adj_method) && (eig.tol >= 0.0) )
    .B <- nrow(M)

    if(!(adj_method %in% c("spectral","nearPD"))){
        stop("in Radj: invalid adj_method")
    }

    if(adj_method=="nearPD" && !is.null(eig.min)){
        stop("in Radj: eig.min applies only with adj_method=='spectral'")
    }

    if(adj_method=="spectral"){
        if(is.null(eig.min)){
            eig.min <- 0.0
        }
        if(eig.tol < eig.min){
            stop("in Radj: eig.tol must be at least as large as eig.min")
        }
    }

    Madj <- M    
    e    <- eigen(M)
    lam  <- e$values
    U    <- e$vectors
    
    if(any(lam < (eig.tol * max(lam)))){
        done <- FALSE
        while(!done){
            if(adj_method=="spectral"){
                lam[which(lam < (eig.tol * max(lam)))] <- eig.min * max(lam)
                tmp <- U %*% diag(lam) %*% t(U)
            } else {
                .res <- nearPD2(Madj, fix0s=TRUE, do2eigen=FALSE, eig.tol=eig.tol, conv.tol=1e-11, maxit=5000)
                if(!.res$converged){
                    stop("in Radj: nearPD2 did not converge; try control$Radj_method = 'spectral'")
                }
                tmp <- as.matrix(.res$mat)
            }
            
            Madj <- sparseMatrix(i = nzlocs[,1], j = nzlocs[,2], x = tmp[nzlocs], dims=c(.B,.B), symmetric=TRUE)
            e    <- eigen(Madj)
            lam  <- e$values
            U    <- e$vectors
            
            done <- all(lam >= (ifelse(adj_method=="spectral", (eig.min * max(lam)), -.Machine$double.eps)))
        }
    }
    return(Madj)
}

Gadj <- function(M, adj_method, eig.tol, eig.min=NULL){
    ## function for G adjustment to PSD, accounting for sum-to-zero constraints
    ##
    ## if adj_method=="nearPD", calls nearPD2 with eig.tol
    ##
    ## if adj_method=="spectral", sets eigenvalues less than (eig.tol * max)
    ## equal to (eig.min * max), aside from the eigenvalue corresponding
    ## to the sum-to-zero constraints
    ##
    ## NOTE: fixed zeros are not allowed, unlike for R adjustment
    stopifnot( (class(M) == "dspMatrix") && is.character(adj_method) && (eig.tol >= 0.0) )
    
    if(!(adj_method %in% c("spectral","nearPD"))){
        stop("in Gadj: invalid adj_method")
    }

    if(adj_method=="nearPD" && !is.null(eig.min)){
        stop("in Gadj: eig.min applies only with adj_method=='spectral'")
    }

    if(adj_method=="spectral"){
        if(is.null(eig.min)){
            eig.min <- 0.0
        }
        if(eig.tol < eig.min){
            stop("in Gadj: eig.tol must be at least as large as eig.min")
        }
    }

    if(max(abs(c(apply(M, 1, sum), apply(M, 2, sum)))) > 1e-10){
        stop("in Gadj: M does not appear to satisfy sum-to-zero constraints")
    }

    Madj <- M
    e    <- eigen(M)
    lam  <- e$values
    U    <- e$vectors
    zloc <- which(apply(U, 2, var) < 1e-10)
    stopifnot(length(zloc)==1L)

    if(any(lam[-zloc] < (eig.tol * max(lam)))){

        if(adj_method=="spectral"){
            tofix      <- setdiff( which(lam < (eig.tol * max(lam))), zloc )
            lam[tofix] <- eig.min * max(lam)
            tmp        <- U %*% diag(lam) %*% t(U)
        } else {
            .res <- nearPD(M, do2eigen=FALSE, eig.tol=eig.tol, conv.tol=1e-11, maxit=5000)
            if(!.res$converged){
                stop("in Gadj: nearPD2 did not converge; try control$Gadj_method = 'spectral'")
            }
            tmp <- as.matrix(.res$mat)
        }

        Madj <- new("dspMatrix", Dim=rep(nrow(tmp),2), x=tmp[lower.tri(tmp, diag=TRUE)], uplo="L")
        if(max(abs(c(apply(Madj, 1, sum), apply(Madj, 2, sum)))) > 1e-10){
            stop("in Gadj: Madj does not satisfy sum-to-zero constraints")
        }
    }
    return(Madj)
}
